### PR TITLE
add types to executeOperation

### PR DESCRIPTION
**Pull request by mmahalwy**
With `executeOperation` becoming the way to test queries and mutations, we lost the ability to specify the returned data and variables needed that existed in apollo-server 2. This adds the ability to add this to `executeOperation` and solve this issue https://github.com/apollographql/apollo-server/issues/5464

Looking for feedback here. I felt like there are one of two ways to tackle this:
- Make this work for `executeOperation` only
- Make `GraphQLRequest` and `GraphQLResponse` generic

Thoughts?

### CODE REVIEW

**Comment by mmahalwy** [comment]
@glasser hey - any feedback on this?

**Comment by glasser** [comment]
Haven't had a chance to review yet!

Are you aware that CI is failing to build it?

**Comment by mmahalwy** [comment]
@glasser yep - aware of CI failing. It's a type error. Easy fix but wanted to make sure we are aligned on implementation first or you truly want `GraphQLRequest` to be a generic

File: packages/apollo-server-core/src/ApolloServer.ts [hunk 1]
```diff
@@ -1,62 +1,64 @@
+import { buildServiceDefinition } from '@apollographql/apollo-tools';
```
Comment (glasser) [inline packages/apollo-server-core/src/ApolloServer.ts:1]: A lot moved around here — did you run some sort of import-sorting script?

File: packages/apollo-server-core/src/ApolloServer.ts [hunk 2]
```diff
@@ -942,12 +944,15 @@ export class ApolloServerBase<
...
    */
-  public async executeOperation(
+  public async executeOperation<
+    TData = Record<string, any>,
+    TVariables = VariableValues,
+  >(
     request: Omit<GraphQLRequest, 'query'> & {
       query?: string | DocumentNode;
-    },
+    } & { variables?: TVariables | VariableValues },
```
Comment (glasser) [inline packages/apollo-server-core/src/ApolloServer.ts:953]: Do we want to add `variables` to the `Omit` above like `query` is?

I don't really understand what the `|` is doing here: why not just TVariables?

File: packages/apollo-server-core/src/ApolloServer.ts [hunk 3]
```diff
@@ -942,12 +944,15 @@ export class ApolloServerBase<
...
     integrationContextArgument?: ContextFunctionParams,
-  ) {
+  ): Promise<GraphQLResponse & { data?: TData | null }> {
```
Comment (glasser) [inline packages/apollo-server-core/src/ApolloServer.ts:955]: Do we want to `Omit` data from GraphQLResponse like we do with query above? Or was that actually a pointless thing to do for query too?

**Review by glasser — CHANGES_REQUESTED** [review]
It would help to see an example of how this code ends up being used.

Is implementing this just for executeOperation an incremental step towards generifying everything or is it a different choice that prevents us from doing that later?

**Comment by mmahalwy** [comment]
@glasser yes. Goal here is being able to type check the variables and output. Usage:

```ts
server.executeOperation<SomeQuery, SomeQueryVariables>()
```

**Comment by glasser** [comment]
Cool. And is the goal to still get TypedDocumentNode working? Will this change be compatible with that?

**Comment by mmahalwy** [comment]
@glasser thats a very good question that I unfortunately don't have an answer to nor have context about what `TypedDocumentNode` is. Rather, my goal was to re-add types back to tests since typing was removed when `apollo-testing-server` was deprecated

**Comment by mmahalwy** [comment]
@glasser bump :)

**Comment by glasser** [comment]
I don't want to change anything about typing for executeOperation that may make it *harder* to later change it to work with TypedDocumentNode.

TypedDocumentNode is a class that code generators like graphql-code-generator know how to generate which both contains the actual parsed document and proper typings for it. So instead of having to manually make sure that the types you put in `<>` match the operation, you let codegen make it impossible to mess up.  https://github.com/dotansimha/graphql-typed-document-node or https://www.graphql-code-generator.com/docs/plugins/typed-document-node for details.

**Comment by mmahalwy** [comment]
closing this :(

**Pull request closed without merging**